### PR TITLE
DefaultBearerTokenResolver additional check for allowUriQueryParameter=false

### PR DESCRIPTION
**Pull request by gburboz**
Updated DefaultBearerTokenResolver to resolve issue  #5668 and specification [rfc6750#section-2](https://tools.ietf.org/html/rfc6750#section-2) compliance

There were significant changes suggested by @jzheaux  in previous PR #6036 hence closed it and raised this new one.

### CODE REVIEW

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 1]
```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
...
 	}
 
-	private static String resolveFromRequestParameters(HttpServletRequest request) {
-		String[] values = request.getParameterValues("access_token");
-		if (values == null || values.length == 0)  {
+	private String resolveFromRequestParameters(final HttpServletRequest request) {
+		// By this point there is either zero or one value
+		final String value = request.getParameter(ACCESS_TOKEN_PARAM_NAME);
+		if (value != null && isParameterTokenSupportedForRequest(request)) {
+			final Matcher matcher = JWT_PATTERN.matcher(value);
```
Comment (jzheaux) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Good catch

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 2]
```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
...
+		} else {
 			return null;
 		}
+	}
 
-		if (values.length == 1) {
-			return values[0];
-		}
+	private boolean isParameterTokenSupportedForRequest(final HttpServletRequest request) {
+		return this.allowUriQueryParameter || (this.allowFormEncodedBodyParameter
```
Comment (vpavic) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: If I'm not missing something, this now allows the use of `allowUriQueryParameter` for all request methods - I don't like that and believe the use of `allowUriQueryParameter` should be limited to `GET` as it originally was. The point is if one can do `POST` then they can provide the parameter as `application/x-www-form-urlencoded` - query parameter support should be only supported as fallback for challenged clients that cannot do anything other than `GET`.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 3]
```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
...
+				&& MediaType.APPLICATION_FORM_URLENCODED.equals(MediaType.valueOf(request.getContentType()))
+				&& !RequestMethod.GET.equals(RequestMethod.valueOf(request.getMethod())));
```
Comment (vpavic) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: If `Content-Type` has been presented then it isn't a `GET` request so this line is needless, right?

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java [hunk 4]
```diff
@@ -43,6 +43,15 @@ public void setUp() {
 		this.resolver = new DefaultBearerTokenResolver();
 	}
 
+	@Test
+	public void resolveWhenNoTokenIsPresentThenTokenIsNotResolved() {
```
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: There is semantic difference that in new test case we have `setAllowFormEncodedBodyParameter(true)` and `setAllowUriQueryParameter(true)` and still it will not resolve the token.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java [hunk 5]
```diff
@@ -128,6 +138,18 @@ public void resolveWhenFormParameterIsPresentAndSupportedThenTokenIsResolved() {
 		assertThat(this.resolver.resolve(request)).isEqualTo(TEST_TOKEN);
 	}
 
+	@Test
+	public void resolveWhenFormParameterIsPresentAndSupportedWithPutRequestThenTokenIsResolved() {
+		this.resolver.setAllowFormEncodedBodyParameter(true);
+
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("PUT");
```
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: Yes, I came across specs while doing this PR and realized the bug. This test case idea is to use cover something other than `GET` and `POST`. Thought about other HTTP methods but then it quickly becomes too many test cases with little value.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java [hunk 6]
```diff
@@ -153,8 +213,22 @@ public void resolveWhenQueryParameterIsPresentAndSupportedThenTokenIsResolved()
...
+	@Test
+	public void resolveWhenQueryParameterIsAbsentAndSupportedWithFormParameterIsPresentAndNotSupportedThenTokenIsResolved() {
+		this.resolver.setAllowUriQueryParameter(true);
+
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("POST");
+		request.setContentType("application/x-www-form-urlencoded");
+		request.addParameter("access_token", TEST_TOKEN);
+
+		assertThat(this.resolver.resolve(request)).isEqualTo(TEST_TOKEN);
```
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: Since form parameter would be more secure than query parameter, I did not add that additional check hence it will resolve. Let me know if this needs to change.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java [hunk 7]
```diff
@@ -586,14 +587,15 @@ public void requestWhenBearerTokenResolverAllowsQueryParameterThenEitherHeaderOr
 
 		JwtDecoder decoder = this.spring.getContext().getBean(JwtDecoder.class);
 		when(decoder.decode(anyString())).thenReturn(JWT);
+		DefaultBearerTokenResolver resolver = this.spring.getContext().getBean(DefaultBearerTokenResolver.class);
+		resolver.setAllowUriQueryParameter(true);
```
[the remark this answers is not in the document]
Comment (gburboz) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java]: That is true! I will make the change to fix this.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 8]
```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
...
+			throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MULTIPLE);
+		}
+		if (this.allowFormEncodedBodyParameter == true) {
+			if (this.allowUriQueryParameter == false && isTokenInUriQueryParameter(request)) {
+				throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MALFORMED);
+			}
+		}
+	}
 
+	private String resolveFromAuthorizationHeader(final HttpServletRequest request) {
```
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Adding `final` allows compiler to do certain optimizations as well also reduces bugs due to accidental changes to value.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 9]
```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
...
-	private static String resolveFromAuthorizationHeader(HttpServletRequest request) {
-		String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
-		if (StringUtils.hasText(authorization) && authorization.startsWith("Bearer")) {
-			Matcher matcher = authorizationPattern.matcher(authorization);
+	private void validate(final HttpServletRequest request, final String authorizationHeaderToken) {
+		if (hasMultipleAccessTokens(request, authorizationHeaderToken)) {
+			throw new OAuth2AuthenticationException(BEARER_TOKEN_ERROR_MULTIPLE);
+		}
+		if (this.allowFormEncodedBodyParameter == true) {
+			if (this.allowUriQueryParameter == false && isTokenInUriQueryParameter(request)) {
```
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Was done like this so that if need be to error out on other conditions for form, we can add them under this

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 2 above)
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Specification does not restrict us to from sending access token in query parameter for POST request

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 10]
```diff
@@ -36,7 +39,21 @@
...
+			BearerTokenErrorCodes.INVALID_REQUEST, HttpStatus.BAD_REQUEST,
+			"Found multiple bearer tokens in the request", "https://tools.ietf.org/html/rfc6750#section-3.1");
+
+	private static final BearerTokenError BEARER_TOKEN_ERROR_MALFORMED = new BearerTokenError(
+			BearerTokenErrorCodes.INVALID_TOKEN, HttpStatus.UNAUTHORIZED, "Bearer token is malformed",
+			"https://tools.ietf.org/html/rfc6750#section-3.1");
+
+	private static final String ACCESS_TOKEN_PARAM_NAME = "access_token";
+
+	private static final String TOKEN_CHARS = "(?<token>([a-zA-Z0-9-._~+/]+)=*$)";
```
Comment (gburboz) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: @vpavic , you are right, use of JWT keyword is incorrect here.

Comment (gburboz) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Earlier regex pattern was incorrect and was excluding `=` chars from extracted token hence updated to new regex, please review

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java [hunk 11]
```diff
@@ -34,7 +34,7 @@
  */
 public class DefaultBearerTokenResolverTests {
 
-	private static final String TEST_TOKEN = "test-token";
+	private static final String TEST_TOKEN = "test-token_+~./dGVzdC10b.2tlbg==";
```
Comment (gburboz) [inline oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: I have updated test token here to include all allowed chars.

**Comment by jzheaux** [comment]
@gburboz, I'd reword your first point a bit and say: Only fix what the task describes. This is typically the ideal. It makes reviews easier and it simplifies backpatching. 

And when folks first start contributing, this is a good rule to follow also for the reason of getting familiar with the contribution guidelines.

There's nothing wrong with wanting to improve other aspects of the surrounding code, and it is certainly welcome, especially if it is a clear win. If the improvements are simple, a separate ticket is usually not necessary (e.g. converting the `BearerTokenError` instances into constants).

For continuity, I recognize that I asked you to split up your commits in such-and-such a way, and I had hoped that such would simplify the review process, but I agree with @vpavic that it is much easier to evaluate small diff, where all of the code in question relates to the ticket in question.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vpavic) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: I don't see how there's a semantic difference when you're not actually passing anything with the request to resolve. This test only verifies that the `DefaultBearerTokenResolver#resolve` returns `null` when there's nothing to resolve (same as `resolveWhenNoHeaderIsPresentThenTokenIsNotResolved`) and that is the same regardless of how you configure `DefaultBearerTokenResolver`.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vpavic) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: If you consider how things are currently implemented then I understand it might appears that other tests for other methods add little value. However, since `POST`, `PUT`, `DELETE` and `PATCH` are the most common verbs used in HTTP based API it makes sense to add them as an insurance to prevent some future change inadvertently breaking those use cases.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vpavic) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: You are making a significant behavior change to `DefaultBearerTokenResolver` here - the point of `allowFormEncodedBodyParameter` switch is to enable form-encoded body parameter and you're making it implicitly enabled via `allowUriQueryParameter` here without that being documented anywhere.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vpavic) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: The only real use for `final` parameters was when dealing with anonymous inner classes and that was removed in Java 8 with introduction of effectively final.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 2 above)
Comment (vpavic) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: The use of query parameter is clearly discouraged by the spec, and it's there just in cases where _it is impossible to transport the access token in the "Authorization" request header field or the HTTP request entity-body_. Having in mind the HTTP semantics, while not explicitly stated, this is clearly targeted at client incapable of doing anything other than a `GET` request. There's no valid reason not to use form-encoded body parameter if client is capable of doing a `POST` request.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java [hunk 12]
```diff
@@ -81,47 +85,65 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
...
+		return multipleParameterTokensPresent || bothParameterTokenAndHeaderPresent;
 	}
 
-	private boolean isParameterTokenSupportedForRequest(HttpServletRequest request) {
-		return ((this.allowFormEncodedBodyParameter && "POST".equals(request.getMethod()))
-				|| (this.allowUriQueryParameter && "GET".equals(request.getMethod())));
+	private boolean isTokenInUriQueryParameter(final HttpServletRequest request) {
+		final String queryString = request.getQueryString();
+		return queryString != null && Arrays.stream(queryString.split("&"))
+				.map(param -> param.split("=")[0])
```
Comment (rwinch) [inline oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: HTTP parameters are URL encoded. This means that technically a URL of `/message?%74oken=abc123` is valid. Also note that using this invoking `HttpServletRequest.getParameter("token")` returns `abc123` We should ensure that we first decode the parameter names before we compare.

**Comment by vpavic** [comment]
If everyone here agrees, I'd open a PR focused purely on addressing #5668 off [this branch](https://github.com/vpavic/spring-security/tree/gh-5668) in my fork. That still leaves room for @gburboz to re-submit the PR with some of the nice optimizations outlined here, like moving exceptions to constants and validating the token chars.

**Comment by rwinch** [comment]
@vpavic Let's let @gburboz clean this up here. 

@gburboz Can you please be sure to address the comments. Please keep the PR focused on this particular issue. If you find more than one issue, please create a separate PR for it. This ensures that users can go through the issues and find out what has changed from version to version. For example, please don't change unrelated code to use final as this is something separate that was changed and (although minor) if we make that change should be included in the change log.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: Both input and expected output is same but this test case gives coverage to different branch logic when form or query parameter is enabled.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: This is how even the existing behavior is. But I do agree with your comment and I will update it as suggested.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: @vpavic  I disagree on real usage you mentioned, as per me real usage is programmer's intent that the value must not be changed (immutability).

I do agree it adds more changes in PR hence additional work to review as well as it may have implications on back patch hence I will revert this change.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Oh! my bad, yes testing a boolean against true or false is lame. Although @vpavic pointed, I didn't realize it until @rwinch pointed it again. I will update it.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: If you still think this is not useful, I will remove this test case once you confirm.

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: @vpavic, **please let me know what you think.**
I was making change as "proposed by @vpavic" to "current test" but realized that "current test" is much similar to as "proposed by @gburboz" (note words Form and Query are only swapped). These last two tests should have similar behaviour is what I am suggesting (could be either `null` or `AuthenticationException`).

- `resolveWhenFormParameterIsPresentAndSupportedWithInvalidContentTypeThenTokenIsNotResolved` (**proposed by @vpavic** )
- `resolveWhenQueryParameterIsAbsentAndSupportedWithFormParameterIsPresentAndNotSupportedThenXyzResult` (**current test**)
- `resolveWhenFormParameterIsAbsentAndSupportedWithQueryParameterIsPresentAndNotSupportedThenAuthenticationExceptionIsThrown` (**proposed by @gburboz** )

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 2 above)
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: It is better to stick to the specifications or we will build solution that is incompatible for some scenarios. For example if we limit only to `GET` method then we are also prohibiting `HEAD` request which is not correct. Let's not add extra restrictions that are not mentioned in specs unless it is required.

I understand "use of query parameter is clearly discouraged by the spec" and that is the reason why we have `allowUriQueryParameter = false` by default.

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 12 above)
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Updated as required

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 2 above)
Comment (jzheaux) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: Agreed that if the client can send a POST, then they can send a POST body. This means that we are within the spec to disallow a query parameter in that case (since it isn't impossible, as @vpavic explained). @gburboz do you have an example where this would not be the case?

This also has the happy consequence that we are not enabling bad behavior, as @rwinch noted.

Also, we can typically relax restrictions iteratively (e.g. in future PRs). And folks can very easily supply their own implementation for their corner case. If it becomes clear that this flexibility is necessary, we can add it with relative ease.

With these items in mind, let's only check for the query parameter when it is a GET request. I like this, too, because it narrows the focus of this PR closer to [its original goal](https://github.com/spring-projects/spring-security/issues/5668).

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: I will remove test case `resolveWhenNoTokenIsPresentThenTokenIsNotResolved()`

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: Added additional `DELETE` and `PATCH` test cases

File: oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolverTests.java]: As proposed by  @vpavic, I have updated test case to `resolveWhenFormParameterIsPresentAndSupportedWithInvalidContentTypeThenTokenIsNotResolved`

File: oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java (hunk 2 above)
Comment (gburboz) [reply oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java]: @jzheaux , below are few examples of query param for method other than `GET` while client is unable to create form body
* `HEAD` method which could be used when `GET` request is processing heavy or results in too much data
* `TRACE` method which can be used for debugging issues with `GET` method
* `POST`/`PUT`/`PATCH`/`DELETE` method but data is sent as query params (non `GET` method used here is to support REST paradigm)

@vpavic since everyone in this PR are leaning towards not supporting scenarios mentioned above, I have reverted the change and limited query param only to `GET` method. In light of above example if we feel change is required, we can raise new issue and raise different PR against it.

**Comment by gburboz** [comment]
@rwinch and @vpavic , I think at this point all the PR comments are resolved. Please let me know if I have missed anything.

**Comment by gburboz** [comment]
> 
> 
> LGTM - I'm going to wait for @vpavic to look it over too

@vpavic, this PR is still waiting on your approval. Unfortunately there seems to be conflicts at this point but I will await your response before resolving them.
